### ISSUE TITLE
loss = nan problem

### ISSUE THREAD

**Issue by sunmiaozju** [MaybeShewill-CV/lanenet-lane-detection#33]
您好，请问我在使用自己制作的数据集进行训练的时候，出现了loss=nan的问题，请问为什么会出现这样的情况呢，一般会是哪里出错呀？
出错信息如下：
![2018-07-25 11-13-06](https://user-images.githubusercontent.com/34623535/43177736-3d8bb556-8ffc-11e8-810a-255721711d74.png)

**Comment by sunmiaozju** [comment]
而且似乎不是特定一组batch图片才会出现nan的问题，因为我把刚才出现nan的那组batch图片提取了出来，由这几张图片（4张）循环往复作为一个新的数据集，用这个新数据集进行训练结果发现又可以正常训练，似乎这组图片又不出现nan的错。

还有一个问题是我发现出现nan的时候，预测的二值化图和实例分割图各个像素都是0.

除此之外，每次训练出现nan的epoch都是不一样的图片batch也不同。

有什么建议吗，您觉得哪里出了问题？@MaybeShewill-CV 感谢

**Comment by MaybeShewill-CV** [comment]
@sunmiaozju 首先确保你的标签正确,其次可能针对不同的数据集你需要自己调整超参数.大多数nan的问题都是由于标签导致的。

**Comment by sunmiaozju** [comment]
我发现在出现loss=nan之前，预测前向输出的binary_logits像素均为0,所以我修改了训练过程，当出现前向输出binary_logits=0的时候，我不进行参数优化而是跳过这个batch，但是我发现之后所有的batch都是一样的出现了问题，即前向输出结果均等于0，。所以我现在认为不是数据集标签的问题，而是模型参数爆炸导致的，具体原因还再找

**Comment by MaybeShewill-CV** [comment]
@sunmiaozju 需要你调整超参数了

**Comment by yiyichun** [comment]
Hi, @sunmiaozju : 
請問你在訓練的時候有修改到那些code嗎? 
我遇到的問題是訓練時的accuracy都是很接近0的值.
![image](https://user-images.githubusercontent.com/34088341/43237692-806c5fc6-90bd-11e8-859f-4bd6cc582db9.png)
感謝你的幫忙 :)

**Comment by yiyichun** [comment]
Hi, @sunmiaozju : 
我剛才試了作者提供的數據集,也就是在data\training_data_example\gt_image_binary 、data\training_data_example\gt_image_instance 、data\training_data_example\image這3個資料夾裡面的6張圖檔,並把它複製成30張,把BATCH_SIZE=2 (用原本的4,記憶體會不夠),但是得到的結果還是一樣耶 ><
![image](https://user-images.githubusercontent.com/34088341/43240008-2e3a34b2-90c7-11e8-9d4a-3f8a80f5ffc5.png)

**Comment by MaybeShewill-CV** [comment]
@yiyichun 等多几个epoch再看看

**Comment by sunmiaozju** [comment]
@MaybeShewill-CV 
您好，我使用tusimple .json文件生成了数据集，并且使用模型训练了起来。但是使用自己的数据集却产生了nan问题。
所以我想问一下，数据集的车道线标签，在聚类的时候或者二值化分割的时候，一定要是连续的车道线吗？我的数据集某些车道线是一个实例类，但是是断开的虚线段。就像下面这样：
![2018-07-26 14-34-04](https://user-images.githubusercontent.com/34623535/43245480-fca50cc8-90e0-11e8-8de8-f370e5ee7f9f.png)

这样的形式会有问题吗？感谢

**Comment by MaybeShewill-CV** [comment]
@sunmiaozju 二值化分割是根据像素点进行分类的

**Comment by sunmiaozju** [comment]
@MaybeShewill-CV 
已经可以训练了，谢谢啊

**Comment by MaybeShewill-CV** [comment]
@yiyichun 你提供的信息不足以定位问题出现的原因, 如可以使用vgg预训练参数作为网络初始值来训练网络而不需要使用--weights_path参数。看这个图的问题你的pixel embedding失效, 导致后期mean shift聚类失败

**Comment by yiyichun** [comment]
hi, @MaybeShewill-CV : 
感謝您的說明. 可能我說的不清楚,我在說明清楚一點好了.
如果是python3 tools/train_lanenet.py --net vgg --dataset_dir data/training_data_example/  ===>accuracy的值不會是0
如果是python3 tools/train_lanenet.py --net vgg --dataset_dir data/training_data_example/ --weights_path model/tusimple_lanenet/tusimple_lanenet_vgg_2018-05-21-11-11-03.ckpt-94000  ===>accuracy值會是0
請問"tusimple_lanenet_vgg_2018-05-21-11-11-03.ckpt-94000"這個是用vgg的網路訓練出來的嗎? 
是否可以提供.pb 檔呢? 
謝謝您

**Comment by south-ocean** [comment]
您好，想问下训练过程中其中的部分epoch出现NAN，您是怎么解决的谢谢！
0727 01:57:11.438254 8959 train_lanenet.py:290] Epoch: 80 total_loss= 1.356155 binary_seg_loss= 0.239498 instance_seg_loss= 3.961687 accuracy= 0.144323 mean_cost_time= 1.340399s 
I0727 01:57:13.380253 8959 train_lanenet.py:290] Epoch: 81 total_loss=    nan binary_seg_loss= 0.229159 instance_seg_loss=    nan accuracy= 0.091045 mean_cost_time= 1.296569s 
I0727 01:57:15.414632 8959 train_lanenet.py:290] Epoch: 82 total_loss= 1.247565 binary_seg_loss= 0.254058 instance_seg_loss= 3.565749 accuracy= 0.105159 mean_cost_time= 1.392229s 
I0727 01:57:17.448236 8959 train_lanenet.py:290] Epoch: 83 total_loss= 1.521073 binary_seg_loss= 0.211007 instance_seg_loss= 4.577894 accuracy= 0.063831 mean_cost_time= 1.382824s 
I0727 01:57:19.482539 8959 train_lanenet.py:290] Epoch: 84 total_loss= 1.441414 binary_seg_loss= 0.211770 instance_seg_loss= 4.310584 accuracy= 0.028835 mean_cost_time= 1.390127s 
I0727 01:57:21.517747 8959 train_lanenet.py:290] Epoch: 85 total_loss= 1.413747 binary_seg_loss= 0.237343 instance_seg_loss= 4.158689 accuracy= 0.072766 mean_cost_time= 1.388580s 
I0727 01:57:23.480197 8959 train_lanenet.py:290] Epoch: 86 total_loss= 1.381895 binary_seg_loss= 0.257384 instance_seg_loss= 4.005753 accuracy= 0.020796 mean_cost_time= 1.394654s 
I0727 01:57:25.489401 8959 train_lanenet.py:290] Epoch: 87 total_loss=    nan binary_seg_loss= 0.184804 instance_seg_loss=    nan accuracy= 0.108197 mean_cost_time= 1.359407s 
@sunmiaozju @MaybeShewill-CV

**Comment by MaybeShewill-CV** [comment]
@yiyichun pb文件你可以用frozen graph脚本自己生成一下哈～

**Comment by MaybeShewill-CV** [comment]
@south-ocean 这个问题最有可能的情况是你的标签像素值不对,　先检查一下标签像素值是否正确吧

**Comment by south-ocean** [comment]
您好，我的label_binary是255像素值，实例分割的为20，70，120，170，220的不同像素值。这样子做的不对么？

**Comment by MaybeShewill-CV** [comment]
@south-ocean 背景是20?

**Comment by MaybeShewill-CV** [comment]
@south-ocean 是的 我的数据集是这样设置的

**Comment by south-ocean** [comment]
好的，我再看下，谢谢！ @MaybeShewill-CV

**Comment by MaybeShewill-CV** [comment]
@south-ocean 好的 欢迎分享结果

**Comment by liuyangly25** [comment]
@MaybeShewill-CV 我的数据也出现了nan的情况，原因是如果label里面是空的（十字路口，停车场，密集车流等）就会产生nan。所以说训练集一定要有车道线了？如果图片中没有车道线，该怎么修改网络呢？

**Comment by MaybeShewill-CV** [comment]
@liuyangly25 放入网络前先检查label 没有车道线的话就跳过这张图像

**Comment by liuyangly25** [comment]
@MaybeShewill-CV 这样的话，测试中一张没有线的照片也会跑出线了。网络能自动判别有没有线吗？

**Comment by MaybeShewill-CV** [comment]
@liuyangly25 没有测试过

**Comment by ding-hai-tao** [comment]
我在训练的时候也遇到nan问题，然后训练停了，我将学习率又降低了10倍后，就可以正常训练了！

**Comment by MaybeShewill-CV** [comment]
@mukaman84 I'll check it whether it is the reason cause the problem. Thanks for sharing it with us.

**Comment by yiyichun** [comment]
@ding-hai-tao , 
你的做法和我的一樣,我也是把學習率降低之後就可以訓練了,但是當epoch跑到一定次數之後,仍會有nan的情形出現.
而且accuracy只有在0.5~0.6之間,請問你有這樣的情形嗎?
謝謝!!

**Comment by ding-hai-tao** [comment]
@yiyichun  目前我才训练到10W次，没有出现nan的情况。

**Comment by ding-hai-tao** [comment]
@yiyichun  我没有改动作者的代码。会不会数据集不一样？

**Comment by yiyichun** [comment]
hi, @ding-hai-tao : 
請問你跑出來的圖會像上面那樣嗎? 我使用的訓練集是TuSimple dataset,裡面的label_data_0313.json、label_data_0531.json、label_data_0601.json裡面的images
謝謝!

**Comment by yiyichun** [comment]
@ding-hai-tao , 感謝你的回復 
請問你跑出來的會長怎樣呢? 
我在想可能是我在製作gt_image_instance、gt_image_binary時有點問題.
請問你是怎樣製作這2個檔案的呢?

**Comment by ding-hai-tao** [comment]
@yiyichun 我是用opencv制作的那两个文档。

**Comment by yiyichun** [comment]
@ding-hai-tao , 
請問方便把你的圖檔寄給我讓我測試看看嗎? 
yiyichun2011@gmail.com 
謝謝!

**Comment by yiyichun** [comment]
@MaybeShewill-CV ,@ding-hai-tao , 
我試出來了,謝謝你們 :) 
目前在還在訓練中,但是有用跑10000次的結果看一下,之前的問題都解掉了.
請問除了tusimple 、culane dataset之外,還有其他的training dataset嗎?

**Comment by MaybeShewill-CV** [comment]
@yiyichun 不客气, 之前是什么原因导致你无法训练的呢?

**Comment by yiyichun** [comment]
@MaybeShewill-CV , 
在讀取gt_image_instance的時候有出現維度錯誤的訊息,我修正維度使之可以訓練,但是訓練出來的結果怪怪的.

**Comment by cardwing** [comment]
@yiyichun , besides, since you mentioned that you have trained your model in CULane dataset, I wonder how well your model performs (e.g., F1 measure in different road categories and in total). I just want to know the performance of the model so that I can decide whether to use this code. Thanks a lot ! ^_^

**Comment by ShangJP** [comment]
@yiyichun @sunmiaozju @south-ocean  请问一下，你们是如何制作那两个文件的啊，我想尝试训练一下作者的代码，第一步就被卡住了，非常感谢，希望能指点一下

**Comment by ding-hai-tao** [comment]
@yiyichun @MaybeShewill-CV  我將学习率降低了十倍后,发现训练得到的结果不好.
![0](https://user-images.githubusercontent.com/33212030/43716559-22309f46-99b7-11e8-9e06-c740c4eb901d.jpg)

![4](https://user-images.githubusercontent.com/33212030/43716609-4c45c752-99b7-11e8-9275-cd16be204cb9.jpg)

这个是过拟合造成的吧?

**Comment by ding-hai-tao** [comment]
@yiyichun 你训练得到的结果好吗?

**Comment by ding-hai-tao** [comment]
@MaybeShewill-CV  训练的时候,什么类型的图片会使训练出现梯度爆炸.

**Comment by cardwing** [comment]
@ding-hai-tao , do you meet the situation where cost, instance cost and binary cost are nan?

**Comment by yiyichun** [comment]
@ding-hai-tao, 我使用約3000張來訓練,約700張來驗證.
train_accuracy約在0.99附近.我跑影片來做測試,有些時候會有抓不到道路資訊.
不過可以增加訓練資料重新訓練,效果會在提升。

**Comment by ding-hai-tao** [comment]
@yiyichun  我用的是70%的图片作为训练集，30%的图片作为验证集。验证集上的acc=0.98+ ，但是小于0.99。用图片测的时候，发现效果不是很好。你的参数是怎么设置的？ 同时我感觉这数据量有点少，你是通过传统的数据增强方法来增加训练资料的嘛？

**Comment by globalmaster** [comment]
> @MaybeShewill-CV
> 已经可以训练了，谢谢啊

你好，我用的是图森的数据集，也出现了nan的情况，请问你是怎么解决的？谢谢

**Comment by phjhuang** [comment]
> @ding-hai-tao, 我使用約3000張來訓練,約700張來驗證.
> train_accuracy約在0.99附近.我跑影片來做測試,有些時候會有抓不到道路資訊.
> 不過可以增加訓練資料重新訓練,效果會在提升。

culane  用来训练效果怎么样？

**Comment by phjhuang** [comment]
@yiyichun   

culane 训练效果怎么样

**Comment by 131404060321** [comment]
@MaybeShewill-CV 我剔除了训练样本中十字路口（gt_label中没有标注）的数据后，程序正常运行没有立即出现 loss = nan 的情况，当我训练一万多次时程序出现了loss=nan的情况，如图：
![loss nan_10245](https://user-images.githubusercontent.com/25121347/46244656-a61b5800-c414-11e8-90cf-ed3ed3210cea.png)
这种情况应该不是数据标注的问题吧，是不是我需要调整一下参数了？

**Comment by MaybeShewill-CV** [comment]
@131404060321 这个应该不是数据标注的问题 应该是需要调整参数

**Comment by MaybeShewill-CV** [comment]
@All The new code updated recently will lead to more stable training process. Welcome
 to test the new code

**Comment by fayechou** [comment]
> @ding-hai-tao ,
> 你的做法和我的一樣,我也是把學習率降低之後就可以訓練了,但是當epoch跑到一定次數之後,仍會有nan的情形出現.
> 而且accuracy只有在0.5~0.6之間,請問你有這樣的情形嗎?
> 謝謝!!

你好，我遇到的问题和你的一模一样，请问你解决了吗？是怎么解决的啊？

**Comment by Sand0001** [comment]
> 已经可以训练了，谢谢啊

请问一下 含有这种不连续的虚线的数据集应该怎么制作instance和binary？是每一类别是一个instance 还是每一段是一个instance 相应的binary应该怎么制作 ，求指导 万分感谢

**Comment by zacario-li** [comment]
请问，我使用tusimple数据集，训练50个epoch后，accuracy直接变成0了，我看了这个issue里面好多人也碰到了这个问题，但是没说是什么原因，请问 @MaybeShewill-CV 你大概知道是哪里有问题吗？

**Comment by ASONG0506** [comment]
> 请问，我使用tusimple数据集，训练50个epoch后，accuracy直接变成0了，我看了这个issue里面好多人也碰到了这个问题，但是没说是什么原因，请问 @MaybeShewill-CV 你大概知道是哪里有问题吗？

请问你解决这个问题了吗

**Comment by MaybeShewill-CV** [comment]
@ASONG0506 你可以检查下你的Tensorflow版本和代码库是否是新的，我本地从来没遇到过这个问题：）

**Comment by niskov** [comment]
> I found which one causes nan. This is because of tf.norm in discriminative_loss_single function. When the smallest value like 0 is given to the input of tf.norm, the derivative of tf.norm is going to inf. So I replaced tf.norm with tf.reduced_sum(tf.square()) and then training is well. Since tf.reduced_sum(tf.square()) isn't identical to tf.norm, Cutoff variance distance and cutoff cluster distance parameters also should be changed. Of course, it depends on the distribution of dataset.

Could you please post your code after the changes?
